compiler/compiler/src/main/java/org/robovm/compiler/AppCompiler.java

```diff
@@ -487,23 +494,158 @@ public void failure(Clazz clazz, Throwable t) {
-        linker.link(linkClasses);
-        long duration = System.currentTimeMillis() - start;
-        config.getLogger().info("Linked %d classes in %.2f seconds", linkClasses.size(), duration / 1000.0);
+        // Persisting the config for later comparison
+        File configFile = new File(config.getTmpDir(), "config.xml");
+        try (FileOutputStream output = new FileOutputStream(configFile)) {
+            try {
+                IOUtils.write(configToXml(config), output, StandardCharsets.UTF_8.toString());
+            } catch (Exception e) {
+                config.getLogger().error("Error when computing config's equality. " +
+                        "Forcing recompilation: %s:%s", e.getClass().getSimpleName(), e.getMessage());
+            }
+        }
+
+    }
+
+    /**
+     * Checks, whether recompilation of the Main binary is necessary by looking at the classPathsFile
+     *
+     * @param classPathsFile
+     * @return
+     * @throws IOException
+     */
+    private boolean needsRecompilation(Config config) throws IOException {
+        if (!config.isSmartSkipRebuild()) {
+            config.getLogger().info("Rebuilding because smartSkipRebuild is disabled. Enable it by adding " +
+                    "<smartSkipRebuild>true</smartSkipRebuild> to your robovm.xml file.");
+            return true;
+        }
+
+        //User can force clean via command line
+        if (config.isClean()) {
+            config.getLogger().info("Config requires clean, rebuilding.");
+            return true;
+        }
+
+        //Check for newer input files compared to Main binary
+        File classPathsFile = new File(config.getTmpDir(), CLASSPATHS_FILENAME);
+        File binaryFile = new File(config.getTmpDir(), "Main");
+
+        if (!classPathsFile.exists()) {
+            config.getLogger().info("Dependency info was not found, compilation is required.");
+            return true;
+        }
+        if(!binaryFile.exists()){
+            config.getLogger().info("Main binary was not found, compilation is required.");
+            return true;
+        }
+
+        //Compare the modification dates of all classes / jars that were linked during the last run with the binary
+        long binaryFileModified = binaryFile.lastModified();
+        try (BufferedReader b = new BufferedReader(new FileReader(classPathsFile))) {
+            String readLine = "";
+            while ((readLine = b.readLine()) != null) {
+                File classPathFile = new File(readLine);
+                // class was removed or is newer?
+                if (!classPathFile.exists() || classPathFile.lastModified() > binaryFileModified) {
+                    config.getLogger().info("Found modified file, compilation is required: %s", classPathFile);
+                    return true;
+                }
+            }
+        }
+
+        //Has the configuration changed between runs (e.g. forcelink)?
+        boolean configsEqual;
```

Comment:
why just don't check timestamps of robovm.xml, robovm.properties and Info.plist.xml files ? 
it will do same but will remove lot of code that was added/changed

Reply:
Creating a dependency to robovm.xml and the like would imply, that you're always configuring RoboVM via these files. This will fall back on you, if you once decide to make a configuration via Gradle or Maven (used today by javafxports and desirable anyways IMHO as of single-source-of-truth principle).
In fact, today you're configuring parts of the build process via the IDEA run configuration (e.g. the architecture - and please don't let us take this information from the build path).
So I think, comparing the configurations this way is required.